File: workflow-left-half-done/node/reconcile-half-run-workflows.js
/**
 * Reconcile Medusa reservations left behind by a workflow that never finished.
 *
 * Medusa v2 workflows are sagas: each step's rollback is opt-in through a compensation
 * function passed to createStep, so a step without one, such as createRemoteLinkStep
 * (GitHub #9844), leaves its side effect in place if a later step throws. Separately, a
 * crashed process or the in-memory Workflow Engine used in production means the saga
 * never reaches the compensating phase at all, so an already-committed reservation from
 * reserveInventoryStep stays committed while workflow_execution is stuck in a
 * non-terminal state (GitHub #9077, #12913, #11266).
 *
 * This lists reservations, resolves each line_item_id's parent order, classifies each
 * reservation with a pure function, and deletes only the unambiguous orphan cases:
 * an order that no longer resolves (404) or an order whose status is canceled.
 * Everything else is reported only. Run on a schedule. Safe to run again and again.
 *
 * Guide: https://www.allanninal.dev/medusa/workflow-left-half-done/
 */
import { pathToFileURL } from "node:url";

const BACKEND_URL = process.env.MEDUSA_BACKEND_URL || "http://localhost:9000";
const ADMIN_EMAIL = process.env.MEDUSA_ADMIN_EMAIL || "admin@example.com";
const ADMIN_PASSWORD = process.env.MEDUSA_ADMIN_PASSWORD || "supersecret";
const STALE_MINUTES = Number(process.env.STALE_MINUTES || 10);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

const DELETABLE = new Set(["orphaned_no_order", "orphaned_canceled_order"]);

/**
 * Pure decision function. No I/O.
 *
 * @param {{ id: string, line_item_id: string | null, created_at: string }} reservation
 * @param {{ id: string, status: string } | null} order
 * @param {string} nowIso
 * @param {number} staleMinutes
 * @returns {"orphaned_no_order" | "orphaned_canceled_order" | "stale_pending_review" | "healthy"}
 */
export function classifyReservation(reservation, order, nowIso, staleMinutes = 10) {
  if (reservation.line_item_id == null) return "healthy";

  if (order == null) return "orphaned_no_order";

  if (order.status === "canceled") return "orphaned_canceled_order";

  const ageMs = Date.parse(nowIso) - Date.parse(reservation.created_at);
  if (ageMs > staleMinutes * 60000 && !["pending", "completed"].includes(order.status)) {
    return "stale_pending_review";
  }

  return "healthy";
}

async function getAdminToken() {
  const res = await fetch(`${BACKEND_URL}/auth/user/emailpass`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ email: ADMIN_EMAIL, password: ADMIN_PASSWORD }),
  });
  if (!res.ok) throw new Error(`Medusa auth ${res.status}`);
  const body = await res.json();
  return body.token;
}

async function adminGet(token, path, params = {}) {
  const url = new URL(`${BACKEND_URL}${path}`);
  for (const [key, value] of Object.entries(params)) {
    url.searchParams.set(key, value);
  }
  const res = await fetch(url, { headers: { Authorization: `Bearer ${token}` } });
  if (!res.ok) throw new Error(`Medusa ${res.status} on GET ${path}`);
  return res.json();
}

async function adminDelete(token, path) {
  const res = await fetch(`${BACKEND_URL}${path}`, {
    method: "DELETE",
    headers: { Authorization: `Bearer ${token}` },
  });
  if (!res.ok) throw new Error(`Medusa ${res.status} on DELETE ${path}`);
  return res.json();
}

async function listReservations(token) {
  const reservations = [];
  let offset = 0;
  const limit = 200;
  while (true) {
    const data = await adminGet(token, "/admin/reservations", {
      fields: "id,line_item_id,inventory_item_id,location_id,quantity,created_at,*line_item.order_id",
      limit,
      offset,
    });
    reservations.push(...data.reservations);
    offset += limit;
    if (offset >= data.count) return reservations;
  }
}

async function getOrderOrNull(token, orderId) {
  if (!orderId) return null;
  const res = await fetch(new URL(`${BACKEND_URL}/admin/orders/${orderId}?fields=id,status`), {
    headers: { Authorization: `Bearer ${token}` },
  });
  if (res.status === 404) return null;
  if (!res.ok) throw new Error(`Medusa ${res.status} on GET /admin/orders/${orderId}`);
  const body = await res.json();
  return body.order;
}

export async function run() {
  const token = await getAdminToken();
  const reservations = await listReservations(token);
  const nowIso = new Date().toISOString();

  let deleted = 0;
  let reported = 0;
  for (const reservation of reservations) {
    const orderId = reservation.line_item?.order_id;
    const order = reservation.line_item_id ? await getOrderOrNull(token, orderId) : null;

    const outcome = classifyReservation(reservation, order, nowIso, STALE_MINUTES);
    if (outcome === "healthy") continue;

    if (DELETABLE.has(outcome)) {
      console.warn(
        `Reservation ${reservation.id} classified as ${outcome}. order_id=${orderId}. ${DRY_RUN ? "Would delete" : "Deleting"}`
      );
      if (!DRY_RUN) await adminDelete(token, `/admin/reservations/${reservation.id}`);
      deleted++;
    } else {
      console.log(
        `Reservation ${reservation.id} reported as ${outcome}. order_id=${orderId} status=${order?.status} (needs human review, not touched).`
      );
      reported++;
    }
  }

  console.log(`Done. ${deleted} reservation(s) ${DRY_RUN ? "to delete" : "deleted"}, ${reported} reservation(s) reported for review.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}

File: workflow-left-half-done/python/reconcile_half_run_workflows.py
"""Reconcile Medusa reservations left behind by a workflow that never finished.

Medusa v2 workflows are sagas: each step's rollback is opt-in through a compensation
function passed to createStep, so a step without one, such as createRemoteLinkStep
(GitHub #9844), leaves its side effect in place if a later step throws. Separately, a
crashed process or the in-memory Workflow Engine used in production means the saga
never reaches the compensating phase at all, so an already-committed reservation from
reserveInventoryStep stays committed while workflow_execution is stuck in a
non-terminal state (GitHub #9077, #12913, #11266).

This lists reservations, resolves each line_item_id's parent order, classifies each
reservation with a pure function, and deletes only the unambiguous orphan cases:
an order that no longer resolves (404) or an order whose status is canceled.
Everything else is reported only. Run on a schedule. Safe to run again and again.
"""
import os
import logging
import requests
from datetime import datetime, timezone

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("reconcile_half_run_workflows")

BACKEND_URL = os.environ["MEDUSA_BACKEND_URL"]
ADMIN_EMAIL = os.environ["MEDUSA_ADMIN_EMAIL"]
ADMIN_PASSWORD = os.environ["MEDUSA_ADMIN_PASSWORD"]
STALE_MINUTES = float(os.environ.get("STALE_MINUTES", "10"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

DELETABLE = {"orphaned_no_order", "orphaned_canceled_order"}


def get_admin_token():
    r = requests.post(
        f"{BACKEND_URL}/auth/user/emailpass",
        json={"email": ADMIN_EMAIL, "password": ADMIN_PASSWORD},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()["token"]


def admin_get(token, path, params=None):
    r = requests.get(
        f"{BACKEND_URL}{path}",
        headers={"Authorization": f"Bearer {token}"},
        params=params or {},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def admin_delete(token, path):
    r = requests.delete(
        f"{BACKEND_URL}{path}",
        headers={"Authorization": f"Bearer {token}"},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def _parse_iso_ms(iso):
    return datetime.fromisoformat(iso.replace("Z", "+00:00")).timestamp() * 1000


def classify_reservation(reservation, order, now_iso, stale_minutes=10):
    """Pure decision function. No I/O.

    reservation: {"id": str, "line_item_id": str | None, "created_at": str (ISO)}
    order: {"id": str, "status": str} | None
    now_iso: str (ISO timestamp)
    stale_minutes: int

    Returns "orphaned_no_order" | "orphaned_canceled_order" |
            "stale_pending_review" | "healthy".
    """
    if reservation.get("line_item_id") is None:
        return "healthy"

    if order is None:
        return "orphaned_no_order"

    if order.get("status") == "canceled":
        return "orphaned_canceled_order"

    age_ms = _parse_iso_ms(now_iso) - _parse_iso_ms(reservation["created_at"])
    if age_ms > stale_minutes * 60000 and order.get("status") not in ("pending", "completed"):
        return "stale_pending_review"

    return "healthy"


def list_reservations(token):
    reservations = []
    offset = 0
    limit = 200
    while True:
        data = admin_get(token, "/admin/reservations", {
            "fields": "id,line_item_id,inventory_item_id,location_id,quantity,created_at,*line_item.order_id",
            "limit": limit,
            "offset": offset,
        })
        reservations.extend(data["reservations"])
        offset += limit
        if offset >= data["count"]:
            return reservations


def get_order_or_none(token, order_id):
    if not order_id:
        return None
    r = requests.get(
        f"{BACKEND_URL}/admin/orders/{order_id}",
        headers={"Authorization": f"Bearer {token}"},
        params={"fields": "id,status"},
        timeout=30,
    )
    if r.status_code == 404:
        return None
    r.raise_for_status()
    return r.json()["order"]


def run():
    token = get_admin_token()
    reservations = list_reservations(token)
    now_iso = datetime.now(timezone.utc).isoformat()

    deleted = 0
    reported = 0
    for reservation in reservations:
        line_item = reservation.get("line_item") or {}
        order_id = line_item.get("order_id")
        order = get_order_or_none(token, order_id) if reservation.get("line_item_id") else None

        outcome = classify_reservation(reservation, order, now_iso, STALE_MINUTES)
        if outcome == "healthy":
            continue

        if outcome in DELETABLE:
            log.warning(
                "Reservation %s classified as %s. order_id=%s. %s",
                reservation["id"], outcome, order_id, "Would delete" if DRY_RUN else "Deleting",
            )
            if not DRY_RUN:
                admin_delete(token, f"/admin/reservations/{reservation['id']}")
            deleted += 1
        else:
            log.info(
                "Reservation %s reported as %s. order_id=%s status=%s (needs human review, not touched).",
                reservation["id"], outcome, order_id, (order or {}).get("status"),
            )
            reported += 1

    log.info(
        "Done. %d reservation(s) %s, %d reservation(s) reported for review.",
        deleted, "to delete" if DRY_RUN else "deleted", reported,
    )


if __name__ == "__main__":
    run()
